x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java

```diff
@@ -675,4 +686,115 @@ protected void parseCreateField(DocumentParserContext context) throws IOExceptio
+            );
+        }
+        return new AggregateMetricSyntheticFieldLoader(name(), simpleName(), metrics);
+    }
+
+    public static class AggregateMetricSyntheticFieldLoader implements SourceLoader.SyntheticFieldLoader {
+        private final String name;
+        private final String simpleName;
+        private final EnumSet<Metric> metrics;
+
+        protected AggregateMetricSyntheticFieldLoader(String name, String simpleName, EnumSet<Metric> metrics) {
+            this.name = name;
+            this.simpleName = simpleName;
+            this.metrics = metrics;
+        }
+
+        @Override
+        public Leaf leaf(LeafReader reader, int[] docIdsInLeaf) throws IOException {
+            Map<Metric, SortedNumericDocValues> metricDocValues = new EnumMap<>(Metric.class);
+            for (Metric m : metrics) {
+                SortedNumericDocValues dv = dv(reader, m);
+                if (dv != null) {
+                    metricDocValues.put(m, dv);
+                }
+            }
+
+            if (metricDocValues.isEmpty()) {
+                return SourceLoader.SyntheticFieldLoader.NOTHING_LEAF;
+            }
+
+            return new AggregateMetricSyntheticFieldLoader.ImmediateLeaf(metricDocValues);
+        }
+
+        private class ImmediateLeaf implements Leaf {
+            private final Map<Metric, SortedNumericDocValues> metricDocValues;
+            private Map<Metric, Boolean> dvHasValue;
+
+            private boolean hasValues;
+
+            ImmediateLeaf(Map<Metric, SortedNumericDocValues> metricDocValues) {
+                this.metricDocValues = metricDocValues;
+            }
+
+            @Override
+            public boolean empty() {
+                return false;
+            }
+
+            @Override
+            public boolean advanceToDoc(int docId) throws IOException {
+                // It is required that all defined metrics must exist. In this case
+                // it is enough to check for the first docValue. However, in the future
+                // we may relax the requirement of all metrics existing. In this case
+                // we should check the doc value for each metric separately
+                dvHasValue = new EnumMap<>(Metric.class);
+                hasValues = false;
+                for (Map.Entry<Metric, SortedNumericDocValues> e : metricDocValues.entrySet()) {
+                    boolean b = e.getValue().advanceExact(docId);
+                    hasValues = hasValues || b;
+                    dvHasValue.put(e.getKey(), b);
+                }
+
+                return hasValues;
+            }
+
+            @Override
+            public void write(XContentBuilder b) throws IOException {
+                if (false == hasValues) {
+                    return;
+                }
+                b.startObject(simpleName);
+                for (Map.Entry<Metric, SortedNumericDocValues> entry : metricDocValues.entrySet()) {
```

Comment:
`metricDocValues` is an `EnumMap` instance. This means that the order that sub fields will be returned in the source, will be the order in the `Metric` enum. Order is`min`,`max`, `sum`, `value_count`. It is not alphabetical, but it will always be consistent. Does it make sense to changes this so sub fields are retuned in an alphabetical order?

Reply:
It's cool. Consistency is lovely for reading these docs. Alphabetical just gets the consistency in an easy way. No need to do it here.